File: rust/src/builder.rs
// Phase 3.7 — MockConfigBuilder: generates synthetic .h5 configs for testing.
// Mirrors Python's MockConfigBuilder (python/src/machine_config/builder.py).
// Values are deterministic/fixed so tests are reproducible across platforms.
// Gaussian correction grids are non-zero so correction-data tests have real
// values to assert against.

use std::path::Path;

use ndarray::Array3;

use crate::error::Result;
use crate::models::*;
use crate::writer::MachineConfigWriter;

const SCHEMA_VERSION: &str = "v1";
const MOCK_EXPORT_DATE: &str = "2026-01-01T00:00:00.000Z";

/// Generates a structurally valid synthetic machine-config `.h5` file.
/// Optical trains have a ClearBox with a non-zero Gaussian correction grid
/// and a placeholder `scan_field_correction_file` dataset.
pub struct MockConfigBuilder {
    pub laser_count: usize,
    pub build_plate_x: f64,
    pub build_plate_y: f64,
    pub build_plate_z: f64,
    pub include_clearbox: bool,
    pub machine_name: String,
    pub manufacturer: String,
    pub model: String,
    pub serial_number: String,
}

impl MockConfigBuilder {
    /// Create a builder for a `laser_count`-laser config with default values.
    pub fn new(laser_count: usize) -> Self {
        Self {
            laser_count,
            build_plate_x: 250.0,
            build_plate_y: 250.0,
            build_plate_z: 20.0,
            include_clearbox: true,
            machine_name: "MockMachine".to_owned(),
            manufacturer: "MockCo".to_owned(),
            model: "MockMIDI+".to_owned(),
            serial_number: "MOCK-001".to_owned(),
        }
    }

    /// Build the in-memory `MachineConfig` without writing to disk.
    pub fn build(&self) -> MachineConfig {
        let meta = MachineConfigMeta {
            schema_version: SCHEMA_VERSION.to_owned(),
            machine_name: self.machine_name.clone(),
            manufacturer: self.manufacturer.clone(),
            model: self.model.clone(),
            serial_number: self.serial_number.clone(),
            file_version: "1.0".to_owned(),
            export_date: MOCK_EXPORT_DATE.to_owned(),
            configuration_hash: "0".repeat(64),
            extra: Default::default(),
        };

        let machine = Machine {
            id: Some("00000000-0000-0000-0000-000000000001".to_owned()),
            machine_name: self.machine_name.clone(),
            manufacturer: self.manufacturer.clone(),
            model: self.model.clone(),
            serial_number: self.serial_number.clone(),
            build_plate_x: Some(self.build_plate_x),
            build_plate_x_unit: Some("mm".to_owned()),
            build_plate_y: Some(self.build_plate_y),
            build_plate_y_unit: Some("mm".to_owned()),
            build_plate_z: Some(self.build_plate_z),
            build_plate_z_unit: Some("mm".to_owned()),
            build_plate_radius: None,
            build_plate_radius_unit: None,
            gas_flow_direction: Some("Y+".to_owned()),
            recoat_direction: Some("X+".to_owned()),
        };

        let optical_trains = (0..self.laser_count)
            .map(|i| self.build_train(i))
            .collect();

        MachineConfig {
            meta,
            machine,
            optical_trains,
            opcua: None,
        }
    }

    /// Build and write the config to `path`.
    pub fn save<P: AsRef<Path>>(&self, path: P) -> Result<()> {
        MachineConfigWriter::new(&self.build()).write(path)
    }

    // -----------------------------------------------------------------------
    // Private
    // -----------------------------------------------------------------------

    fn build_train(&self, index: usize) -> OpticalTrain {
        let sign = if index % 2 == 0 { -1.0_f64 } else { 1.0_f64 };

        let scanner = Scanner {
            manufacturer: "MockCo".to_owned(),
            model: "MockScan".to_owned(),
            serial_number: format!("MOCK-SC-{:02}", index + 1),
            working_distance: Some(670.0),
            working_distance_unit: Some("mm".to_owned()),
            scan_field_x: Some(600.0),
            scan_field_x_unit: Some("mm".to_owned()),
            scan_field_y: Some(600.0),
            scan_field_y_unit: Some("mm".to_owned()),
            scan_field_z: Some(76.5),
            scan_field_z_unit: Some("mm".to_owned()),
            scan_head_offset_x: Some(sign * 87.5),
            scan_head_offset_x_unit: Some("mm".to_owned()),
            scan_head_offset_y: Some(sign * -23.5),
            scan_head_offset_y_unit: Some("mm".to_owned()),
            scan_head_offset_z: Some(-1.0),
            scan_head_offset_z_unit: Some("mm".to_owned()),
            scan_head_rotation: Some(if index % 2 == 0 { 0.0 } else { 180.0 }),
            scan_head_rotation_unit: Some("degrees".to_owned()),
            axis_configuration: Some("3D".to_owned()),
            x_axis: mock_axis(),
            y_axis: mock_axis(),
            z_axis: Some(mock_axis()),
            focus: None,
        };

        let light_source = LightSource {
            manufacturer: "MockLaser".to_owned(),
            model: "MockFiber-1070".to_owned(),
            serial_number: format!("MOCK-LS-{:02}", index + 1),
            wavelength: Some(1070.0),
            wavelength_unit: Some("nm".to_owned()),
            power_max_nominal: Some(1000.0),
            power_max_nominal_unit: Some("W".to_owned()),
            power_max_actual: Some(1020.0),
            power_max_actual_unit: Some("W".to_owned()),
            power_min_actual: Some(100.0),
            power_min_actual_unit: Some("W".to_owned()),
            power_min_nominal: None,
            power_min_nominal_unit: Some("W".to_owned()),
            power_bit_resolution: None,
            power_bit_resolution_unit: Some("bits".to_owned()),
            watts_to_volts_algorithm: Some("LINEAR".to_owned()),
            watts_to_volts_params: Some("[1,100,10,1000]".to_owned()),
        };

        let collimator = Collimator {
            manufacturer: "MockOptics".to_owned(),
            model: "D50_F120".to_owned(),
            serial_number: format!("MOCK-COL-{:02}", index + 1),
            focal_length: Some(120.0),
            focal_length_unit: Some("mm".to_owned()),
        };

        let scanner_card = ScannerCard {
            manufacturer: "Raylase".to_owned(),
            model: "SP-ICE-3".to_owned(),
            serial_number: format!("MOCK-SC-CARD-{:02}", index + 1),
            communication_protocol: Some("SL2-100".to_owned()),
            sample_period: Some(10.0),
            sample_period_unit: Some("μs".to_owned()),
        };

        OpticalTrain {
            train_id: format!("Optical_Train_{:02}", index + 1),
            id: None,
            beam_profile_type: None,
            beam_waist_definition: Some("knife-edge".to_owned()),
            beam_waist_major: Some(67.0),
            beam_waist_major_unit: Some("μm".to_owned()),
            beam_waist_minor: Some(68.0),
            beam_waist_minor_unit: Some("μm".to_owned()),
            beam_waist_offset_z: Some(0.5),
            beam_waist_offset_z_unit: Some("mm".to_owned()),
            build_plane_offset_major: Some(0.2),
            build_plane_offset_major_unit: Some("mm".to_owned()),
            build_plane_offset_minor: Some(0.7),
            build_plane_offset_minor_unit: Some("mm".to_owned()),
            collimator_focal_length: Some(120.0),
            collimator_focal_length_unit: Some("mm".to_owned()),
            m2_major: Some(1.05),
            m2_minor: Some(1.08),
            major_axis_angle: Some(0.0),
            major_axis_angle_unit: Some("degrees".to_owned()),
            rayleigh_length_major: Some(3.1),
            rayleigh_length_major_unit: Some("mm".to_owned()),
            rayleigh_length_minor: Some(3.2),
            rayleigh_length_minor_unit: Some("mm".to_owned()),
            scanner_number: None,
            thermal_lensing_passed: Some(false),
            thermal_lensing_focal_plane_shift: Some(1.0),
            thermal_lensing_focal_plane_shift_unit: Some("mm".to_owned()),
            thermal_lensing_threshold: Some(0.75),
            thermal_lensing_threshold_unit: Some("mm".to_owned()),
            scanner,
            light_source,
            collimator,
            scanner_card,
            clearbox: if self.include_clearbox { Some(mock_clearbox(index)) } else { None },
            scan_field_correction_file: if self.include_clearbox { Some(mock_sfcf(index)) } else { None },
        }
    }
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

fn mock_axis() -> AxisConfig {
    AxisConfig {
        actual_bit_resolution: Some(20),
        actual_bit_resolution_unit: Some("bits".to_owned()),
        commanded_bit_resolution: Some(20),
        commanded_bit_resolution_unit: Some("bits".to_owned()),
        control_type: None,
        range_of_motion: None,
        range_of_motion_unit: Some("mm".to_owned()),
        smoothing_kernel: Some("GAUSSIAN".to_owned()),
        smoothing_parameters: Some(60.0),
        tuning_parameters: None,
        tuning_type: None,
    }
}

fn mock_clearbox(index: usize) -> ClearBox {
    let grid = gaussian_correction_grid();
    let mut inv = grid.clone();
    inv.mapv_inplace(|v| v * 0.9);
    ClearBox {
        ip_address: format!("192.168.1.{}", 10 + index),
        serial_number: Some(format!("{:03}", index + 1)),
        data_port: Some(5001),
        server_port: Some(20101),
        actual_timing_offset: Some(-8),
        commanded_timing_offset: Some(50),
        correction_data: Some(array3_to_nested(&grid)),
        inverse_correction_data: Some(array3_to_nested(&inv)),
        manufacturer: None,
        model: None,
        output_path: Some("/recordings/".to_owned()),
        selected_camera: Some("Default".to_owned()),
        custom_video_format: Some("MP4".to_owned()),
        video_output: Some("HDMI".to_owned()),
        show_console: Some(false),
        software_trigger_delay: Some(3000),
        volts_to_watts_algorithm: Some("LINEAR".to_owned()),
        volts_to_watts_params: Some("50.0,100.0".to_owned()),
        correction_grid_domain_shape: None,
        inverse_grid_domain_shape: None,
    }
}

fn mock_sfcf(index: usize) -> ScanFieldCorrectionFile {
    ScanFieldCorrectionFile {
        document_name: format!("mock_laser_{}.fc3", index + 1),
        document_id: format!("00000000-0000-0000-0000-{:012}", index + 1),
        file_size: 1024,
        valid_as_of_date: MOCK_EXPORT_DATE.to_owned(),
        document_created_at: None,
        document_type: Some("Scan Field Correction File".to_owned()),
        original_uri: None,
        raw_bytes: None,
    }
}

/// Smooth 2-D Gaussian warp pattern, shape (257, 257, 2).
/// Non-zero by design — zeros would hide correction-application bugs.
fn gaussian_correction_grid() -> Array3<f64> {
    let n = 257_usize;
    let mut grid = Array3::<f64>::zeros((n, n, 2));
    for i in 0..n {
        let x = -1.0_f64 + 2.0 * i as f64 / (n - 1) as f64;
        for j in 0..n {
            let y = -1.0_f64 + 2.0 * j as f64 / (n - 1) as f64;
            let warp = 2.0_f64 * (-(x * x + y * y) / 0.5_f64).exp();
            grid[[i, j, 0]] = warp;
            grid[[i, j, 1]] = warp * 0.8;
        }
    }
    grid
}

/// Convert a `(D0, D1, D2)` array to a nested `Vec` with NaN mapped to `None`.
fn array3_to_nested(arr: &Array3<f64>) -> Vec<Vec<Vec<Option<f64>>>> {
    let shape = arr.shape();
    (0..shape[0])
        .map(|i| {
            (0..shape[1])
                .map(|j| {
                    (0..shape[2])
                        .map(|k| {
                            let v = arr[[i, j, k]];
                            if v.is_nan() { None } else { Some(v) }
                        })
                        .collect()
                })
                .collect()
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::reader::MachineConfigReader;

    #[test]
    fn builder_1_laser_roundtrip() {
        let tmp = tempfile();
        MockConfigBuilder::new(1).save(&tmp).unwrap();
        let config = MachineConfigReader::open(&tmp).unwrap().parse().unwrap();
        assert_eq!(config.optical_trains.len(), 1);
        assert_eq!(config.meta.machine_name, "MockMachine");
    }

    #[test]
    fn builder_2_laser_roundtrip() {
        let tmp = tempfile();
        MockConfigBuilder::new(2).save(&tmp).unwrap();
        let config = MachineConfigReader::open(&tmp).unwrap().parse().unwrap();
        assert_eq!(config.optical_trains.len(), 2);
    }

    #[test]
    fn builder_plate_dimensions() {
        let tmp = tempfile();
        MockConfigBuilder::new(1).save(&tmp).unwrap();
        let config = MachineConfigReader::open(&tmp).unwrap().parse().unwrap();
        assert_eq!(config.machine.build_plate_x, Some(250.0));
        assert_eq!(config.machine.build_plate_y, Some(250.0));
        assert_eq!(config.machine.build_plate_z, Some(20.0));
    }

    #[test]
    fn builder_correction_grid_shape() {
        let tmp = tempfile();
        MockConfigBuilder::new(1).save(&tmp).unwrap();
        let cd = MachineConfigReader::open(&tmp).unwrap().get_correction_data(0).unwrap();
        assert_eq!(cd.shape, [257, 257, 2]);
    }

    #[test]
    fn builder_correction_grid_nonzero() {
        let tmp = tempfile();
        MockConfigBuilder::new(1).save(&tmp).unwrap();
        let cd = MachineConfigReader::open(&tmp).unwrap().get_correction_data(0).unwrap();
        // Centre cell (128, 128, 0) must be the Gaussian peak (warp ≈ 2.0 for r=0).
        // Row-major index: i * d1 * d2 + j * d2 + k  where shape = [257, 257, 2].
        let centre = cd.data[128 * 257 * 2 + 128 * 2];
        assert!(centre > 1.9 && centre < 2.1, "centre={centre}");
        assert!(cd.data.iter().any(|v| v.abs() > 0.0), "all zeros");
    }

    #[test]
    fn builder_no_clearbox() {
        let tmp = tempfile();
        let mut b = MockConfigBuilder::new(1);
        b.include_clearbox = false;
        b.save(&tmp).unwrap();
        let config = MachineConfigReader::open(&tmp).unwrap().parse().unwrap();
        assert!(config.optical_trains[0].clearbox.is_none());
    }

    // Minimal temp-file helper for inline tests — avoids a `tempfile` crate
    // dependency in the lib target (it is a dev-dependency already from
    // Cargo.toml, but we can't use it in `#[cfg(test)]` inside lib without
    // adding it to regular deps, so we use a deterministic path in the OS
    // temp dir instead).
    fn tempfile() -> std::path::PathBuf {
        let mut p = std::env::temp_dir();
        // Each call gets a unique name via a thread-local counter.
        use std::sync::atomic::{AtomicU32, Ordering};
        static COUNTER: AtomicU32 = AtomicU32::new(0);
        p.push(format!("mc_builder_test_{}.h5", COUNTER.fetch_add(1, Ordering::Relaxed)));
        p
    }
}

File: go/builder.go
package machineconfig

import (
	"fmt"
	"math"

	"machine-config-go/capabilities/v1_0/layout"
)

const (
	mockExportDate    = "2026-01-01T00:00:00.000Z"
	mockSchemaVersion = "v1"
)

// MockConfigBuilder generates deterministic synthetic machine-config files for testing.
// Values are fixed so tests are reproducible across platforms.
type MockConfigBuilder struct {
	NLasers         int
	MachineName     string
	BuildPlateX     float64
	BuildPlateY     float64
	IncludeClearbox bool
}

// NewMockConfigBuilder returns a builder with sensible defaults.
func NewMockConfigBuilder() *MockConfigBuilder {
	return &MockConfigBuilder{
		NLasers:         2,
		MachineName:     "MockMachine",
		BuildPlateX:     250.0,
		BuildPlateY:     250.0,
		IncludeClearbox: true,
	}
}

// Build constructs the in-memory MachineConfig without writing to disk.
func (b *MockConfigBuilder) Build() *MachineConfig {
	meta := MachineConfigMeta{
		SchemaVersion:     mockSchemaVersion,
		MachineName:       b.MachineName,
		Manufacturer:      "MockCo",
		Model:             "MockMIDI+",
		SerialNumber:      "MOCK-001",
		FileVersion:       "1.0",
		ExportDate:        mockExportDate,
		ConfigurationHash: "0000000000000000000000000000000000000000000000000000000000000000",
		Extra:             map[string]any{},
	}

	machineID := "00000000-0000-0000-0000-000000000001"
	machine := Machine{
		ID:              StrPtr(machineID),
		MachineName:     b.MachineName,
		Manufacturer:    "MockCo",
		Model:           "MockMIDI+",
		SerialNumber:    "MOCK-001",
		BuildPlateX:     Float64Ptr(b.BuildPlateX),
		BuildPlateXUnit: StrPtr("mm"),
		BuildPlateY:     Float64Ptr(b.BuildPlateY),
		BuildPlateYUnit: StrPtr("mm"),
		BuildPlateZ:     Float64Ptr(20.0),
		BuildPlateZUnit: StrPtr("mm"),
		GasFlowDirection: StrPtr("Y+"),
		RecoatDirection:  StrPtr("X+"),
	}

	trains := make([]OpticalTrain, b.NLasers)
	for i := range trains {
		trains[i] = b.buildTrain(i)
	}

	return &MachineConfig{
		Meta:          meta,
		Machine:       machine,
		OpticalTrains: trains,
		Opcua:         nil,
	}
}

// Save builds the config and writes it to path.
func (b *MockConfigBuilder) Save(path string) error {
	return NewWriter().Write(b.Build(), path)
}

func (b *MockConfigBuilder) buildTrain(index int) OpticalTrain {
	sign := -1.0
	if index%2 != 0 {
		sign = 1.0
	}
	rotation := 0.0
	if index%2 != 0 {
		rotation = 180.0
	}

	ax := mockAxis()
	scanner := Scanner{
		Manufacturer:          "MockCo",
		Model:                 "MockScan",
		SerialNumber:          fmt.Sprintf("MOCK-SC-%02d", index+1),
		WorkingDistance:       Float64Ptr(670.0),
		WorkingDistanceUnit:   StrPtr("mm"),
		ScanFieldX:            Float64Ptr(600.0),
		ScanFieldXUnit:        StrPtr("mm"),
		ScanFieldY:            Float64Ptr(600.0),
		ScanFieldYUnit:        StrPtr("mm"),
		ScanFieldZ:            Float64Ptr(76.5),
		ScanFieldZUnit:        StrPtr("mm"),
		ScanHeadOffsetX:       Float64Ptr(sign * 87.5),
		ScanHeadOffsetXUnit:   StrPtr("mm"),
		ScanHeadOffsetY:       Float64Ptr(sign * -23.5),
		ScanHeadOffsetYUnit:   StrPtr("mm"),
		ScanHeadOffsetZ:       Float64Ptr(-1.0),
		ScanHeadOffsetZUnit:   StrPtr("mm"),
		ScanHeadRotation:      Float64Ptr(rotation),
		ScanHeadRotationUnit:  StrPtr("degrees"),
		AxisConfiguration:     StrPtr("3D"),
		XAxis:                 ax,
		YAxis:                 ax,
		ZAxis:                 &ax,
		Focus:                 nil,
	}

	ls := LightSource{
		Manufacturer:          "MockLaser",
		Model:                 "MockFiber-1070",
		SerialNumber:          fmt.Sprintf("MOCK-LS-%02d", index+1),
		Wavelength:            Float64Ptr(1070.0),
		WavelengthUnit:        StrPtr("nm"),
		PowerMaxNominal:       Float64Ptr(1000.0),
		PowerMaxNominalUnit:   StrPtr("W"),
		PowerMaxActual:        Float64Ptr(1020.0),
		PowerMaxActualUnit:    StrPtr("W"),
		PowerMinActual:        Float64Ptr(100.0),
		PowerMinActualUnit:    StrPtr("W"),
		PowerMinNominalUnit:   StrPtr("W"),
		PowerBitResolutionUnit: StrPtr("bits"),
		WattsToVoltsAlgorithm: StrPtr("LINEAR"),
		WattsToVoltsParams:    StrPtr("[1,100,10,1000]"),
	}

	col := Collimator{
		Manufacturer:    "MockOptics",
		Model:           "D50_F120",
		SerialNumber:    fmt.Sprintf("MOCK-COL-%02d", index+1),
		FocalLength:     Float64Ptr(120.0),
		FocalLengthUnit: StrPtr("mm"),
	}

	sc := ScannerCard{
		Manufacturer:          "Raylase",
		Model:                 "SP-ICE-3",
		SerialNumber:          fmt.Sprintf("MOCK-SC-CARD-%02d", index+1),
		CommunicationProtocol: StrPtr("SL2-100"),
		SamplePeriod:          Float64Ptr(10.0),
		SamplePeriodUnit:      StrPtr("μs"),
	}

	var cb *ClearBox
	var sfcf *ScanFieldCorrectionFile
	if b.IncludeClearbox {
		c := mockClearbox(index)
		cb = &c
		s := mockSFCF(index)
		sfcf = &s
	}

	return OpticalTrain{
		TrainID:                          layout.TrainID(index),
		BeamWaistDefinition:              StrPtr("knife-edge"),
		BeamWaistMajor:                   Float64Ptr(67.0),
		BeamWaistMajorUnit:               StrPtr("μm"),
		BeamWaistMinor:                   Float64Ptr(68.0),
		BeamWaistMinorUnit:               StrPtr("μm"),
		BeamWaistOffsetZ:                 Float64Ptr(0.5),
		BeamWaistOffsetZUnit:             StrPtr("mm"),
		BuildPlaneOffsetMajor:            Float64Ptr(0.2),
		BuildPlaneOffsetMajorUnit:        StrPtr("mm"),
		BuildPlaneOffsetMinor:            Float64Ptr(0.7),
		BuildPlaneOffsetMinorUnit:        StrPtr("mm"),
		CollimatorFocalLength:            Float64Ptr(120.0),
		CollimatorFocalLengthUnit:        StrPtr("mm"),
		M2Major:                          Float64Ptr(1.05),
		M2Minor:                          Float64Ptr(1.08),
		MajorAxisAngle:                   Float64Ptr(0.0),
		MajorAxisAngleUnit:               StrPtr("degrees"),
		RayleighLengthMajor:              Float64Ptr(3.1),
		RayleighLengthMajorUnit:          StrPtr("mm"),
		RayleighLengthMinor:              Float64Ptr(3.2),
		RayleighLengthMinorUnit:          StrPtr("mm"),
		ThermalLensingPassed:             BoolPtr(false),
		ThermalLensingFocalPlaneShift:    Float64Ptr(1.0),
		ThermalLensingFocalPlaneShiftUnit: StrPtr("mm"),
		ThermalLensingThreshold:          Float64Ptr(0.75),
		ThermalLensingThresholdUnit:      StrPtr("mm"),
		Scanner:                          scanner,
		LightSource:                      ls,
		Collimator:                       col,
		ScannerCard:                      sc,
		OptionalComponents:               OptionalComponents{Clearbox: cb},
		ScanFieldCorrectionFile:          sfcf,
	}
}

func mockAxis() AxisConfig {
	return AxisConfig{
		ActualBitResolution:         IntPtr(20),
		ActualBitResolutionUnit:     StrPtr("bits"),
		CommandedBitResolution:      IntPtr(20),
		CommandedBitResolutionUnit:  StrPtr("bits"),
		SmoothingKernel:             StrPtr("GAUSSIAN"),
		SmoothingParameters:         Float64Ptr(60.0),
		RangeOfMotionUnit:           StrPtr("mm"),
	}
}

func mockClearbox(index int) ClearBox {
	fwd := gaussianCorrectionGrid()
	inv := scaleGrid(fwd, 0.9)
	return ClearBox{
		IPAddress:             fmt.Sprintf("192.168.1.%d", 10+index),
		SerialNumber:          StrPtr(fmt.Sprintf("%03d", index+1)),
		DataPort:              IntPtr(5001),
		ServerPort:            IntPtr(20101),
		ActualTimingOffset:    IntPtr(-8),
		CommandedTimingOffset: IntPtr(50),
		CorrectionData:        &fwd,
		InverseCorrectionData: &inv,
		OutputPath:            StrPtr("/recordings/"),
		SelectedCamera:        StrPtr("Default"),
		CustomVideoFormat:     StrPtr("MP4"),
		VideoOutput:           StrPtr("HDMI"),
		ShowConsole:           BoolPtr(false),
		SoftwareTriggerDelay:  IntPtr(3000),
		VoltsToWattsAlgorithm: StrPtr("LINEAR"),
		VoltsToWattsParams:    StrPtr("50.0,100.0"),
	}
}

func mockSFCF(index int) ScanFieldCorrectionFile {
	return ScanFieldCorrectionFile{
		DocumentName:  fmt.Sprintf("mock_laser_%d.fc3", index+1),
		DocumentID:    fmt.Sprintf("00000000-0000-0000-0000-%012d", index+1),
		FileSize:      1024,
		ValidAsOfDate: mockExportDate,
		DocumentType:  StrPtr("Scan Field Correction File"),
	}
}

// gaussianCorrectionGrid builds a 257×257×2 Gaussian warp pattern.
// Matches the Rust builder exactly: x/y normalized to [-1,1], warp = 2*exp(-(x²+y²)/0.5).
func gaussianCorrectionGrid() [][][]*float64 {
	const n = 257
	grid := make([][][]*float64, n)
	for i := 0; i < n; i++ {
		x := -1.0 + 2.0*float64(i)/float64(n-1)
		grid[i] = make([][]*float64, n)
		for j := 0; j < n; j++ {
			y := -1.0 + 2.0*float64(j)/float64(n-1)
			warp := 2.0 * math.Exp(-(x*x+y*y)/0.5)
			v0, v1 := warp, warp*0.8
			grid[i][j] = []*float64{&v0, &v1}
		}
	}
	return grid
}

// scaleGrid multiplies every non-nil cell by factor.
func scaleGrid(src [][][]*float64, factor float64) [][][]*float64 {
	dst := make([][][]*float64, len(src))
	for i, row := range src {
		dst[i] = make([][]*float64, len(row))
		for j, col := range row {
			dst[i][j] = make([]*float64, len(col))
			for k, cell := range col {
				if cell == nil {
					dst[i][j][k] = nil
				} else {
					v := *cell * factor
					dst[i][j][k] = &v
				}
			}
		}
	}
	return dst
}
